File: src/main/java/com/xander/sword/App03.java
package com.xander.sword;

import java.util.ArrayList;

/**
 * 
 * 输入一个链表，按链表从尾到头的顺序返回一个ArrayList。
 * 
 */

public class App03 {

    public static class ListNode {
        int val;
        ListNode next = null;
        ListNode(int val) {
            this.val = val;
        }
    }

    public static void main(String[] args) {
        App03 app = new App03();
        ListNode head = new ListNode(1);
        ListNode pre = head;
        for (int i = 2; i < 9; i++) {
            ListNode tmp = new ListNode(i);
            pre.next = tmp;
            pre = pre.next;
        }
        ArrayList<Integer> res = app.printListFromTailToHead(head);
        System.out.println(res);
    }

    public ArrayList<Integer> printListFromTailToHead(ListNode listNode) {
        ListNode head = new ListNode(0);
        while (listNode != null) {
            ListNode headNext = head.next;
            head.next = listNode;
            listNode = listNode.next;
            head.next.next = headNext;
        }
        ArrayList<Integer> res = new ArrayList<>();
        ListNode p = head.next;
        while (p != null) {
            res.add(p.val);
            p = p.next;
        }
        return res;
    }
}

File: src/main/kotlin/com/xander/sword/k/App03.kt
package com.xander.sword.k

/**
 *
 * 输入一个链表，按链表从尾到头的顺序返回一个ArrayList。
 */
object App03 {

    fun turnOverLinkedList(head: Node<String>): ArrayList<String> {
        // 先翻转
        var listSize = 0
        var node = head.next
        head.next = null
        while (node != null) {
            // println("1 node: $node")
            val tmp = node
            node = node.next
            val tail = head.next
            head.next = tmp
            tmp.next = tail
            listSize++
        }
        // 然后构建
        val list = ArrayList<String>(listSize)
        node = head.next
        while (null != node) {
            // println("2 node: $node")
            list.add(node.value)
            node = node.next
        }
        return list
    }

    class Node<T>(var value: T) {
        var next: Node<T>? = null

        fun link(nextValue: T): Node<T> {
            if (null == next) {
                next = Node(nextValue)
            } else {
                next?.value = nextValue
            }
            return next!!
        }

        override fun toString(): String {
            return value.toString()
        }
    }

}


fun main() {
    val head = App03.Node<String>("")
    head.link("9")
        .link("8")
        .link("7")
        .link("6")
        .link("5")
        .link("4")
        .link("3")
        .link("2")
        .link("1")
    val list = App03.turnOverLinkedList(head)
    list.forEach {
        print("$it ")
    }
}
